pudb/theme.py

```diff
@@ -23,1023 +23,856 @@
 
-        "var label": ("dark blue", "dark cyan"),
-        "var value": ("black", "dark cyan"),
-        "focused var label": ("dark blue", "dark green"),
-        "focused var value": ("black", "dark green"),
+    "focused breakpoint": "focused sidebar two",
+    "focused current breakpoint": "focused breakpoint",
+    "focused disabled breakpoint": "focused sidebar three",
+    "focused disabled current breakpoint": "focused disabled breakpoint",
+    # }}}
 
-        "highlighted var label": ("white", "dark cyan"),
-        "highlighted var value": ("black", "dark cyan"),
-        "focused highlighted var label": ("white", "dark green"),
-        "focused highlighted var value": ("black", "dark green"),
+    # {{{ shell
+    "command line edit": "source",
+    "command line output": "source",
+    "command line prompt": "source",
+    "command line input": "source",
+    "command line error": "warning",
 
-        "return label": ("white", "dark blue"),
-        "return value": ("black", "dark cyan"),
-        "focused return label": ("light gray", "dark blue"),
-        "focused return value": ("black", "dark green"),
+    "focused command line output": "focused source",
+    "focused command line input": "focused source",
+    "focused command line error": "focused source",
 
-        # }}}
+    "command line clear button": "button",
+    "command line focused button": "focused button",
+    # }}}
 
-        # {{{ stack
+    # {{{ Code syntax
+    "comment":      "source",
+    "keyword":      "source",
+    "literal":      "source",
+    "name":         "source",
+    "operator":     "source",
+    "punctuation":  "source",
+    "argument":     "name",
+    "builtin":      "name",
+    "exception":    "name",
+    "function":     "name",
+    "pseudo":       "builtin",
+    "class":        "function",
+    "dunder":       "function",
+    "magic":        "dunder",
+    "namespace":    "keyword",
+    "keyword2":     "keyword",
+    "string":       "literal",
+    "doublestring": "string",
+    "singlestring": "string",
+    "docstring":    "string",
+    "backtick":     "string",
+    # }}}
+}
+INHERITANCE_MAP = CLEAN_INHERITANCE_MAP.copy()
+
+
+def get_style(palette_dict: dict, style_name: str):
+    """
+    Recursively search up the style hierarchy for the first style which has
+    been defined, and add it to the palette_dict under the given style_name.
+    """
+    try:
+        style = palette_dict[style_name]
+        if not isinstance(style, PaletteEntry):
+            style = PaletteEntry(style_name, *style)
+            style.handle_256_colors()
+            palette_dict[style_name] = style
+        return style
+    except KeyError:
+        parent_name = INHERITANCE_MAP[style_name]
+        style = replace(get_style(palette_dict, parent_name), name=style_name)
+        palette_dict[style_name] = style
+        return style
+
+
+def link(child: str, parent: str):
+    INHERITANCE_MAP[child] = parent
```

Comment:
I'm not a huge fan of the globally stateful design in use here. (I.e. I'm not liking that things are being modified in-place, and I like even less that these things are global variables.) Could this be refactored so that nothing needs to be modified in place? I think that might be possible by having each theme provide an additional inheritance map that's "overlaid" over the default one. But I'd be happy to consider other designs as well.

Reply:
Yeah sure! This approach was a kludge that I threw in when I realized that themes could interfere with each other. I like the idea of having a local map overlaid over the global default one, that should be easy to do.

Reply:
Okay, I added a local dict of inheritance overrides that is looked up first, and which is local to `get_palette`